rust/codelist-rs/src/codelist_factory.rs

```diff
@@ -0,0 +1,915 @@
+use crate::errors::CodeListError;
+use crate::codelist::CodeList;
+use crate::codelist::CodeListOptions;
+use crate::metadata::{Metadata, MetadataSource};
+use crate::types::CodeListType;
+
+/// Struct to represent a codelist factory, which is used to load codelists from a directory and make sure all codelists are created following the same rules
+///
+/// # Fields
+/// * `codelist_options` - The options for the codelist
+/// * `metadata` - The metadata for the codelist
+/// * `codelist_type` - The type of codelist
+pub struct CodeListFactory {
+    pub codelist_options: CodeListOptions,
+    pub metadata: Metadata,
+    pub codelist_type: CodeListType,
+}
+
+impl CodeListFactory {
+    /// Create a new codelist factory
+    ///
+    /// # Arguments
+    /// * `codelist_options` - The options for the codelist
+    /// * `metadata` - The metadata for the codelist
+    /// * `codelist_type` - The type of codelist
+    pub fn new(codelist_options: CodeListOptions, metadata: Metadata, codelist_type: CodeListType) -> Self {
+        CodeListFactory {
+            codelist_options: codelist_options.clone(),
+            metadata: metadata.clone(),
+            codelist_type: codelist_type.clone(),
+        }
+    }
+
+    /// Load a codelist from a csv file
+    ///
+    /// # Arguments
+    /// * `file_path` - The path to the csv file
+    /// 
+    /// # Returns
+    /// * `Result<CodeList, CodeListError>` - The codelist or an error
+    /// 
+    /// # Errors
+    /// * `CodeListError::IOError` - If there is an error reading the file
+    /// * `CodeListError::CSVError` - If there is an error parsing the CSV file
+    /// * `CodeListError::InvalidCodeField` - If the code field is missing from the JSON object
+    /// * `CodeListError::InvalidTermField` - If the term field is missing from the JSON object
+    /// * `CodeListError::InvalidCode` - If the code value is not a valid string
+    /// * `CodeListError::InvalidTerm` - If the term value is not a valid string
+    /// * `CodeListError::EmptyCode` - If the code value is an empty string
+    /// * `CodeListError::EmptyTerm` - If the term value is an empty string
```

Comment:
I think we can just shorten this to 

```rs
    /// # Returns
    /// * `Result<CodeList, CodeListError>` - The codelist or an error
    ```
    
    only. i love docs but i think in this case it is slightly hard to read and means if we change or add any error types we have to change them here and in the error file docs/